### DISCUSSION TITLE
Removal of file ownership checks in our internal web server

### DISCUSSION THREAD

**Discussion by Ferroin** [netdata/netdata#12253] [Ideas]
Currently, our internal web server performs it’s own checks of file ownership independently of the normal UNIX DAC checks performed by the OS, and refuses to serve the files. Officially, these checks exist to prevent the web server from serving arbitrary files from the host system.

However, these checks have their own set of associated issues:

* They don’t actually adequately protect against the very category of attacks they are intended to protect against.
* Even if they did provide adequate protection, users who needed this protection would already be getting such protection through other means because they would need it for all applications, not just Netdata.
* They regularly cause confusion for users, more so than any other aspect of our internal web server.
* Our current usage actually _enables_ a whole different class of attacks, and avoiding those attacks would either eliminate the limited protections provided by the checks or significantly complicate our handling of users and groups.

Given all of these issues, I propose that we remove these checks and just let the OS do it’s job managing file access.

## Inadequate protection against serving arbitrary files.

There are, in general, three situations where the agent may be coerced into serving arbitrary local files:

* A malicious local user with admin access moves files on the server _or_ modifies the Netdata configuration. These checks by definition _cannot_ protect against this type of attack, as a user capable of making such an attack can either modify the config so the checks pass, or modify file ownership so that the checks pass.
* A malicious local or remote user utilizes a bug that allows for manipulating the path passed to the `open()` call in he web server. The checks do protect against this case, but it is also significantly less likely than the first possibility.
* A malicious local or remote user utilizes a bug that allows for arbitrary code execution within the execution context of our built-in web server. These checks _cannot_ protect against this type of attack, because by definition the attacker can simply bypass them.

## Not a required line of defense for a majority of users.

Put simply, the type of attack these checks try to prevent is not a normal concern for a majority of users. Serving arbitrary files from a given system is rarely, if ever, a goal for an attacker unless they know there is sensitive data on the system _and_ they have n interest in exfiltrating it.

For users for which that is actually the case, they will almost certainly already be using some form of protection beyond the basic UNIX DAC permissions model, such as SELinux, Apparmor, or even use of hardware-level security mechanisms such as a dedicated HSM.

For users for which that is _not_ the case, these checks provide no practical benefit (especially since some of the things that might be targets for such an attack on a system without sensitive data are already available simply by having access to the dashboard).

## A regular source of confusion for users.

Over the past 1.5 years, we have had more than a dozen instances of users reporting issues that were ultimately determined to be caused by these checks, and all of them were functionally cases of things working as intended, but not being configured correctly.

In comparison, no other issues reported involving our integrated web server have cropped up more than once over the past year and a half, and all of them have been actual bugs in our code.

This is a solid indication that the functionally hidden additional permissions requirements are not intuitive to our users, and thus will _always_ be a potential source of confusion.

## Enabler for a whole separate class of attacks.

Put simply, our current usage of these checks involves, by default, requiring the files to be owned by the `netdata` user and the `netdata` group. Combined with the default file permissions, this means that _the agent web server has write permissions on the dashboard files_.

This, given that we provide no way for remote users to modify these files, is in and of itself a potentially serious security issue, as it would allow an attacker who gets arbitrary code execution in the context of the internal web server to replace dashboard files with malicious code.

We can _technically_ resolve this particular issue without removing these checks, but doing so would either reduce the (already ineffective) protections they provide or significantly complicate our handling of users and groups for Netdata.

**Comment by ilyam8** [comment]
@ktsaou please check this

**Comment by ktsaou** [comment]
I understand these checks have introduced a lot of issues and since netdata already checks the path and does not allow servicing files outside its web folder, I agree to remove these checks.

**Reply by ktsaou** [reply]
@cronlabspl you don't want these checks to be removed? Why? What is wrong?

**Reply by Ferroin** [reply]
@cronlabspl

Simply trying to open a file automatically checks file permissions, it’s an inherent part of the `open()` system call. Netdata currently _additionally_ calls `stat()` on each file it would serve with it’s web sever and compares the owner and group to what it thinks they should be before choosing to open the file in question.

This is about removing those additional checks, which are not actually necessary (because the OS is _already_ checking that we have appropriate permissions to access the file before letting us do so).

Note that this would bring us in line with how pretty much every web server in existence (Apache, NGinx, Lightspeed, Cherokee, etc) works, they just trust the OS to handle permissions correctly.

**Reply by ktsaou** [reply]
> will be able to upload **any** file

Also, the is no code in netdata to upload any file.

In addition to what @Ferroin said, netdata also has strict requirements about the directory files are being served from. The checks to be removed are additional, on top of what the O/S does and the directory checks. In simple words, we will not check for a specific owner and group on the web files we serve. We will continue to check the directory and let the O/S make sure user `netdata` can read them.

**Reply by Ferroin** [reply]
> @Ferroin
> 
> > Netdata currently additionally calls stat() on each file it would serve with it’s web sever and compares the owner and group to what it thinks they should be before choosing to open the file in question.
> 
> This other check should be done by OS, not additional app like ND.

Exactly, which is why this discussion here originated about removing those checks.

**Comment by ktsaou** [comment]
Thank you for bringing this up again @Ferroin !
